packages/google_sign_in/lib/google_sign_in_tizen.dart

```diff
@@ -0,0 +1,302 @@
+    List<String> scopes = const <String>[],
+    SignInOption signInOption = SignInOption.standard,
+    String? hostedDomain,
+    String? clientId,
+  }) async {
+    if (signInOption == SignInOption.games) {
+      throw PlatformException(
+        code: 'unsupported-options',
+        message: 'Games sign in is not supported on Tizen.',
+      );
+    }
+
+    if (clientId != null) {
+      throw PlatformException(
+        code: 'invalid-parameter',
+        message:
+            'ClientID cannot be set when initializing GoogleSignIn for Tizen, '
+            'instead call `setCredentials` in google_sign_in_tizen.dart.',
+      );
+    }
+
+    _ensureInitParametersInitialized();
+    _initParameters = _SignInInitParametersTizen(
+      clientId: _initParameters!.clientId,
+      clientSecret: _initParameters!.clientSecret,
+      scopes: scopes,
+    );
+  }
+
+  @override
+  Future<GoogleSignInUserData?> signInSilently() {
+    throw UnimplementedError('signInSilently() has not been implemented.');
+  }
+
+  @override
+  Future<GoogleSignInUserData?> signIn() async {
+    _ensureInitParametersInitialized();
+    _ensureNavigatorKeyInitialized();
+
+    if (_tokenData != null) {
+      return _createUserData(_tokenData!.idToken);
+    }
+
+    final AuthorizationResponse authorizationResponse =
+        await _authClient.requestAuthorization(
+      _initParameters!.clientId,
+      _initParameters!.scopes,
+    );
+
+    final Future<TokenResponse?> tokenResponseFuture = _authClient.pollToken(
+      clientId: _initParameters!.clientId,
+      clientSecret: _initParameters!.clientSecret,
+      deviceCode: authorizationResponse.deviceCode,
+      interval: authorizationResponse.interval,
+    );
+
+    device_flow_widget.showDeviceFlowWidget(
+      code: authorizationResponse.userCode,
+      verificationUrl: authorizationResponse.verificationUrl,
+      expiresIn: authorizationResponse.expiresIn,
+      onExpired: () => _authClient.cancelPollToken(),
+      onCancelled: () => _authClient.cancelPollToken(),
+    );
+
+    // Waits until user interaction on secondary device is finished, or until
+    // code is expired or polling is cancelled.
+    final TokenResponse? tokenResponse = await tokenResponseFuture;
+    if (tokenResponse == null) {
+      return null;
+    }
+    device_flow_widget.closeDeviceFlowWidget();
+
+    _tokenData = _GoogleSignInTokenDataTizen(
+      accessToken: tokenResponse.accessToken,
+      accessTokenExpirationDate: DateTime.now().add(tokenResponse.expiresIn),
+      refreshToken: tokenResponse.refreshToken,
+      idToken: tokenResponse.idToken,
+    );
+
+    return _createUserData(_tokenData!.idToken);
+  }
+
+  @override
+  Future<GoogleSignInTokenData> getTokens(
+      {required String email, bool? shouldRecoverAuth = true}) async {
+    if (_tokenData == null) {
+      throw PlatformException(
+          code: 'not-signed-in',
+          message: 'Cannot get tokens as there are no signed in user.');
```

Comment:
For Android, [GoogleAuthUtil.getToken](https://developers.google.com/android/reference/com/google/android/gms/auth/GoogleAuthUtil#getToken(android.content.Context,%20android.accounts.Account,%20java.lang.String,%20android.os.Bundle)) is called, the API tries to access token from signed-in user. If that's not possible, [UserRecoverableAuthException](https://developers.google.com/android/reference/com/google/android/gms/auth/UserRecoverableAuthException) will be thrown. In the case of `google_sign_in_android`, it catches this exception and starts interactive sign-in automatically: https://github.com/flutter/plugins/blob/9a1749b65de6966eb94d402a5ab1ace5eaacc18c/packages/google_sign_in/google_sign_in_android/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java#L626-L663.

For iOS, [GIDAuthentication.doWithFreshTokens](https://github.com/google/GoogleSignIn-iOS/blob/b945c693d0a7fbfc32ec5a56b5d5cdc4561730df/GoogleSignIn/Sources/GIDAuthentication.m#L222) is called. Based on [google_sign_in_ios](https://github.com/flutter/plugins/blob/9a1749b65de6966eb94d402a5ab1ace5eaacc18c/packages/google_sign_in/google_sign_in_ios/ios/Classes/FLTGoogleSignInPlugin.m#L150-L152) and, I believe method will never return as

```objc
result(error != nil ? getFlutterError(error) : @{
  @"idToken" : authentication.idToken,
  @"accessToken" : authentication.accessToken,
});
```

will not be called when auth is `nil`. (Maybe this is a bug? I'm not quite familiar with Objective-C.)

Reply:
Hmm, I don't know what's the difference between `REQUEST_CODE_SIGNIN` and `REQUEST_CODE_RECOVER_AUTH` (in the Android implementation you linked) but your current code should be enough. Thank you for the detailed analysis.

Also `are`→`is`.

Reply:
They act as enums and are passed to `startActivityForResult` ([REQUEST_CODE_SIGNIN](https://github.com/flutter/plugins/blob/9a1749b65de6966eb94d402a5ab1ace5eaacc18c/packages/google_sign_in/google_sign_in_android/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java#L428) and [REQUEST_CODE_RECOVER_AUTH](https://github.com/flutter/plugins/blob/9a1749b65de6966eb94d402a5ab1ace5eaacc18c/packages/google_sign_in/google_sign_in_android/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java#L650)), which are again passed to [onActivityResult](https://github.com/flutter/plugins/blob/9a1749b65de6966eb94d402a5ab1ace5eaacc18c/packages/google_sign_in/google_sign_in_android/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java#L667) (in `requestCode`). The method `onActivityResult` checks this value to determine why it was called and handle return data differently. In the case of `REQUEST_CODE_RECOVER_AUTH`, the method will return a token object while in the case of `REQUEST_CODE_SIGNIN`, it will return a `GoogleSignInAccount` object. 

> `are` -> `is`.

Right, thanks!